sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -1043,6 +1045,34 @@ object TypeCoercion {
       }
     }
   }
+
+  /**
+   * A special rule to support string literal as the second argument of date_add/date_sub functions,
+   * to keep backward compatibility as a temporary workaround.
+   * TODO(SPARK-28589): implement ANSI type type coercion and handle string literals.
+   */
+  object StringLiteralCoercion extends TypeCoercionRule {
```

Comment:
This causes a behavior difference in arithmatic operations, too. Could you describe the following change in the PR description?

**2.4.5 and 3.0.0-preview2**
```
scala> sql("select (cast('2020-03-28' AS DATE) + '1')").show
org.apache.spark.sql.AnalysisException: cannot resolve '(CAST('2020-03-28' AS DATE) + CAST('1' AS DOUBLE))' due to data type mismatch: differing types in '(CAST('2020-03-28' AS DATE) + CAST('1' AS DOUBLE))' (date and double).; line 1 pos 8;
```

This PR.
```
scala> sql("select (cast('2020-03-28' AS DATE) + '1')").show
+-------------------------------------+
|date_add(CAST(2020-03-28 AS DATE), 1)|
+-------------------------------------+
|                           2020-03-29|
+-------------------------------------+
```

Reply:
You forgot to write a function name in the second query above?
```
scala> sql("select (cast('2020-03-28' AS DATE) + '1')").show
                 ^^^^
```

Reply:
No~ That's the behavior of this PR. This PR extends accidentally *expressions`, too.

Reply:
Oh, I see.